internal/graph/lookupsubjects.go

```diff
@@ -0,0 +1,495 @@
+		log.Ctx(ctx).Trace().Msg("exclusion")
+		return cl.lookupSetOperation(ctx, req, rw.Exclusion, newLookupSubjectsExclusion(stream))
+	default:
+		return fmt.Errorf("unknown kind of rewrite in lookup subjects")
+	}
+}
+
+func (cl *ConcurrentLookupSubjects) lookupSetOperation(
+	ctx context.Context,
+	req ValidatedLookupSubjectsRequest,
+	so *core.SetOperation,
+	reducer lookupSubjectsReducer,
+) error {
+	cancelCtx, checkCancel := context.WithCancel(ctx)
+	defer checkCancel()
+
+	g, subCtx := errgroup.WithContext(cancelCtx)
+	g.SetLimit(int(cl.concurrencyLimit))
+
+	for index, childOneof := range so.Child {
+		stream := reducer.ForIndex(subCtx, index)
+
+		switch child := childOneof.ChildType.(type) {
+		case *core.SetOperation_Child_XThis:
+			return errors.New("use of _this is unsupported; please rewrite your schema")
+
+		case *core.SetOperation_Child_ComputedUserset:
+			g.Go(func() error {
+				return cl.lookupViaComputed(subCtx, req, stream, child.ComputedUserset)
+			})
+
+		case *core.SetOperation_Child_UsersetRewrite:
+			g.Go(func() error {
+				return cl.lookupViaRewrite(subCtx, req, stream, child.UsersetRewrite)
+			})
+
+		case *core.SetOperation_Child_TupleToUserset:
+			g.Go(func() error {
+				return cl.lookupViaTupleToUserset(subCtx, req, stream, child.TupleToUserset)
+			})
+
+		case *core.SetOperation_Child_XNil:
+			// Purposely do nothing.
+			continue
+
+		default:
+			return fmt.Errorf("unknown set operation child `%T` in expand", child)
+		}
+	}
+
+	// Wait for all dispatched operations to complete.
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	return reducer.CompletedChildOperations()
+}
+
+const dispatchChunkSize = 100
+
+func (cl *ConcurrentLookupSubjects) dispatchTo(
+	ctx context.Context,
+	parentRequest ValidatedLookupSubjectsRequest,
+	toDispatchByType *tuple.ONRByTypeSet,
+	parentStream dispatch.LookupSubjectsStream,
+) error {
+	if toDispatchByType.IsEmpty() {
+		return nil
+	}
+
+	cancelCtx, checkCancel := context.WithCancel(ctx)
+	defer checkCancel()
+
+	g, subCtx := errgroup.WithContext(cancelCtx)
+	g.SetLimit(int(cl.concurrencyLimit))
+
+	stream := &dispatch.WrappedDispatchStream[*v1.DispatchLookupSubjectsResponse]{
+		Stream: parentStream,
+		Ctx:    subCtx,
+		Processor: func(result *v1.DispatchLookupSubjectsResponse) (*v1.DispatchLookupSubjectsResponse, bool, error) {
+			return &v1.DispatchLookupSubjectsResponse{
+				FoundSubjectIds: result.FoundSubjectIds,
+				Metadata:        addCallToResponseMetadata(result.Metadata),
+			}, true, nil
+		},
+	}
+
+	toDispatchByType.ForEachType(func(resourceType *core.RelationReference, resourceIds []string) {
+		chunkCount := (len(resourceIds) / dispatchChunkSize) + 1
+		for chunkIndex := 0; chunkIndex < chunkCount; chunkIndex++ {
+			chunkStart := chunkIndex * dispatchChunkSize
+			chunkEnd := (chunkIndex + 1) * dispatchChunkSize
+			if chunkEnd > len(resourceIds) {
+				chunkEnd = len(resourceIds)
+			}
+
+			chunkIds := resourceIds[chunkStart:chunkEnd]
```

Comment:
Now that Go has generics, it might useful to make this a generic function for batching up slices.
Reading the code that sets up indexes is noisy and error-prone. It'd be great if we only had to write it once.